src/storage/src/hummock/state_store.rs

```diff
@@ -194,74 +194,84 @@ impl HummockStorage {
+
+            // check epoch validity
+            validate_epoch(read_version.pinned_version.safe_epoch(), epoch)?;
+
+            // Query shared buffer. Return the value without iterating SSTs if found
+            for shared_buffer in read_version.shared_buffer {
+                for batch in shared_buffer.get_overlap_batches(&(key..=key), false) {
+                    match batch.get(key) {
+                        Some(v) => {
+                            self.stats.get_shared_buffer_hit_counts.inc();
+                            return Ok(v.into_user_value().map(|v| v.into()));
                         }
+                        None => continue,
                     }
                 }
+            }
 
-                (read_version.uncommitted_ssts, read_version.pinned_version)
-            };
+            (read_version.uncommitted_ssts, read_version.pinned_version)
+        };
 
-            let mut table_counts = 0;
-            let internal_key = key_with_epoch(key.to_vec(), epoch);
+        let mut table_counts = 0;
+        let internal_key = key_with_epoch(key.to_vec(), epoch);
 
-            // Query uploaded but uncommitted SSTs. Return the value if found.
-            let table_infos = prune_ssts(uncommitted_ssts.iter(), &(key..=key), false);
-            for table_info in table_infos.into_iter().rev() {
-                let table = self.sstable_store.sstable(table_info.id).await?;
-                table_counts += 1;
-                if let Some(v) = self.get_from_table(table, &internal_key, key).await? {
-                    return Ok(Some(v));
-                }
+        // Query uploaded but uncommitted SSTs. Return the value if found.
+        let table_infos = prune_ssts(uncommitted_ssts.iter(), &(key..=key), false);
+        for table_info in table_infos.into_iter().rev() {
+            let table = self.sstable_store.sstable(table_info.id).await?;
+            table_counts += 1;
+            if let Some(v) = self.get_from_table(table, &internal_key, key).await? {
+                return Ok(Some(v));
             }
+        }
 
-            for level in pinned_version.levels() {
-                if level.table_infos.is_empty() {
-                    continue;
-                }
-                match level.level_type() {
-                    LevelType::Overlapping => {
-                        let table_infos = prune_ssts(level.table_infos.iter(), &(key..=key), false);
-                        for table_info in table_infos.into_iter().rev() {
+        for level in pinned_version.levels() {
+            if level.table_infos.is_empty() {
+                continue;
+            }
+            match level.level_type() {
+                LevelType::Overlapping => {
+                    let table_infos = prune_ssts(level.table_infos.iter(), &(key..=key), false);
+                    for table_info in table_infos.into_iter().rev() {
+                        if !vnode_bitmap_filter(
```

Comment:
Using vnode bitmap to filter SSTs is a sst pruning strategy. Can we put this check in `prune_ssts` and only return ssts that pass the check?

Reply:
> Using vnode bitmap to filter SSTs is a sst pruning strategy. Can we put this check in `prune_ssts` and only return ssts that pass the check?

but parameter of `prune_ssts` is range of key, which could include lots of different vnode hashval

Reply:
True. `prune_ssts` should filter SSTs by a set for vnodes. I think executor will also pass a set of vnodes it is interested at to scan/iter interfaces so we can use the same `prune_ssts` to do SST pruning.

Reply:
> True. `prune_ssts` should filter SSTs by a set for vnodes. I think executor will also pass a set of vnodes it is interested at to scan/iter interfaces so we can use the same `prune_ssts` to do SST pruning.

done